### DISCUSSION TITLE
realtime subscription works in dev env but not on production

### DISCUSSION THREAD

**Discussion by mikanesth** [pocketbase/pocketbase#1550] [Q&A] [answered]
Hello everybody,

I work on a SvelteKit app that require realtime updates on certain pages.
I implemented PocketBase with the 'SSR way' described by the author of this great package ;-) and it works flawlessly for Auth, every CRUD operation and loading data for the page.
If I reload the page I got, thanks to the load() function, the updated version of the page with the updated data from the pocketable collection and it works perfectly. But I would like to get live updates thanks to the pocketable realtime subscription.

I managed to make it works in dev environment like this 

```
//+page.svelte
<script>
	import { serializePOJO } from '$lib/serialize';
	import { onMount } from 'svelte';

     /** @type {import('./$types').PageData} */
     export let data;
    $: ({ dataSalles } = data);

onMount(()=>{
        const pb = new PocketBase('http://127.0.0.1:8090')
        async function update(){
            const updateData = serializePOJO(await pb.collection("my_collection").getFullList());
            data.dataSalles = updateData
        };
        pb.realtime.subscribe('my_collection', function(e){
            console.log(e.record);
            update()
        })
    })
</script>
```

It works perfectly well with the vite dev server but as soon as I build or preview the production version the live reload of the data on the page doesn't works anymore. I can refresh and get the updated data though.

In PROD env if I go to the network debug panel I can see that the realtime request is made and pocketBase responses with 200 for the first GET method and 204 for the POST one. So should be OK?

This error is reproductible if I deploy a node version of the app or a serverless version on Vercel or cloudfalre Pages. 

Before posting I obviously take several hours to get it working and reads docs and posts as much as I can.
I know that there is a problem with the EventSource with node, but I wait for onMount() so the code should execute on client side with browser loaded.

I hope this is clear don't hesitate if you need any information, thanks for your help

**Comment by mikanesth** [comment]
ps:
I also tried 
```
async function update(){
            const updateData = serializePOJO(await pb.collection("my_collection").getFullList());
            data.dataSalles = updateData
        };
pb.collection('my_collection').subscribe('*', update)
```

instead of
```
pb.realtime.subscribe('my_collection', function(e){
            console.log(e.record);
            update()
        })
```
Idem: Works in DEV not in PROD

**Comment by ganigeorgiev** [comment, marked as answer]
Is there any error thrown in the node console or browser?

`subscribe` returns a promise so you can try to `catch` and inspect it, eg:
```js
pb.collection('salles_bloc').subscribe('*', function(e){
    console.log(e.record);
    update()
}).catch((err) => {
    console.log(err.originalError);
})
```

You can also try starting PocketBase with `--debug` and check realtime logs output in the console.

Also please note that in the above examples you are not using the SSR PocketBase instance and you'll receive event messages only if your request client satisfy the `salles_bloc` List API rule.
The SvelteKit SSR examples in https://github.com/pocketbase/js-sdk#ssr-integration works only within a server-side context. If you want a mixed approach, you'll have to load your cookie via `document.cookie` in a separate client-side instance.

**Comment by mikanesth** [comment]
Thanks for your answer,

Will try  and come back.

I thought that the request made on the client side, will use the AuthStore created by the server side registration.
It works in dev.
So I will also try to load a client-side instance with loadFromCookie()

Thank you

**Reply by ganigeorgiev** [reply]
> I thought that the request made on the client side, will use the AuthStore created by the server side registration. It works in dev.

No, it shouldn't work. You can't "share" complex js objects over the network, they'll need to be serialized somehow (that's also the reason why you need to return POJO from server-side load functions). If it works on `dev`, then something else is happening but I'm not sure what.

**Reply by mikanesth** [reply]
Sorry for wasting your time

**Comment by mikanesth** [comment]
OK big mistake from me
It works when I create a client side instance

But I still do not understand why it works on the Vite dev server with my garbage first version

Thanks a lot !!!!!

**Reply by jamauro** [reply]
I'm exploring the same in SvelteKit. Curious how you got it all working. Would you mind sharing what you ended up doing?

**Reply by mikanesth** [reply]
Hello,
My problem was that I misunderstood what's going on between server and client side, thinking that once an authStore was generated on the server side I could invoke pb.realtime method on the client side without creating a valid authStore on the client side. 
I was confused by the dev server of SvelteKit that make it works despite the fact that I was all wrong.

I have done exactly what Gani proposed namely generating a PocketBase instance on th client side using the cookie.
For doing so your cookie cannot be "html only" because you have to access it in your javascript.

On the client side it looks like this 

```javascript
<script>
    import PocketBase from 'pocketbase';
    import { onDestroy, onMount } from 'svelte';
    /** @type {import('./$types').PageData} */
    export let data;

    onMount(async ()=>{
        const pb = new PocketBase('https://xxxxxx.com');
        pb.authStore.loadFromCookie(document.cookie || '')
        async function update(){
            const updateData = await pb.collection('azerty').getFirstListItem(`first_definition= first thing`))
            data.something = updateData
        }
        pb.realtime.subscribe('azerty', update) 
    });
    onDestroy(()=>{
        const pb = new PocketBase('https://xxxxx.com');
        pb.authStore.clear()
    })
</script>
```

If it can be of any help

**Reply by mikanesth** [reply]
the onDestroy() part of the code is up to you it's not mandatory. PocketBase create the authStore on the browser local storage and you can use it through all of your application if you want

**Reply by jamauro** [reply]
Thanks!
